### PR TITLE
Add support for Arm FHM

### PR DESCRIPTION
**Pull request by kulinseth**

### CODE REVIEW

**Comment by kulinseth** [comment]
@malfet ping for review.

**Review by malfet — APPROVED** [review]

**Review by malfet — APPROVED** [review]

File: src/arm/mach/init.c [hunk 1]
```diff
@@ -376,6 +376,17 @@ void cpuinfo_arm_mach_init(void) {
 		cpuinfo_isa.i8mm = true;
 	}
 
+	const uint32_t has_FEAT_FHM = get_sys_info_by_name("hw.optional.arm.FEAT_FHM");
```
Comment (Maratyszcza) [inline src/arm/mach/init.c:379]: Use lower-case characters in identifiers, i.e. `has_feat_fhm`

File: src/arm/mach/init.c [hunk 2]
```diff
@@ -376,6 +376,17 @@ void cpuinfo_arm_mach_init(void) {
...
+	if (has_FEAT_FHM != 0) {
+		cpuinfo_isa.fhm = true;
+	} else {
+		// Prior to iOS 15, use 'hw.optional.armv8_2_fhm'
+		const uint32_t has_FEAT_FHM_legacy = get_sys_info_by_name("hw.optional.armv8_2_fhm");
```
Comment (Maratyszcza) [inline src/arm/mach/init.c:384]: Ditto

File: include/cpuinfo.h [hunk 3]
```diff
@@ -1759,6 +1760,14 @@ static inline bool cpuinfo_has_arm_fcma(void) {
...
 }
 
+static inline bool cpuinfo_has_arm_fhm(void) {
+	#if CPUINFO_ARCH_ARM || CPUINFO_ARCH_ARM64
+		return cpuinfo_isa.fhm;
+	#else
+		return false;
+	#endif
+}
+
```
Comment (Maratyszcza) [inline include/cpuinfo.h:1770]: Move right after `cpuinfo_has_arm_fp16_arith` as these two are related.

File: tools/isa-info.c [hunk 4]
```diff
@@ -133,6 +133,7 @@ int main(int argc, char** argv) {
 		printf("\tVFPv4: %s\n", cpuinfo_has_arm_vfpv4() ? "yes" : "no");
 		printf("\tVFPv4+D32: %s\n", cpuinfo_has_arm_vfpv4_d32() ? "yes" : "no");
 		printf("\tVJCVT: %s\n", cpuinfo_has_arm_jscvt() ? "yes" : "no");
+		printf("\tFMLAL/FMLSL: %s\n", cpuinfo_has_arm_fhm() ? "yes" : "no");
```
Comment (Maratyszcza) [inline tools/isa-info.c:136]: Don't report it under "Floating-Point" section because these are SIMD-only instructions

File: tools/isa-info.c [hunk 5]
```diff
@@ -158,6 +160,7 @@ int main(int argc, char** argv) {
 		printf("\tARM v8.1 SQRDMLxH: %s\n", cpuinfo_has_arm_neon_rdm() ? "yes" : "no");
 		printf("\tARM v8.2 FP16 arithmetics: %s\n", cpuinfo_has_arm_fp16_arith() ? "yes" : "no");
 		printf("\tARM v8.2 BF16: %s\n", cpuinfo_has_arm_bf16() ? "yes" : "no");
+		printf("\tARM v8.2 FHM: %s\n", cpuinfo_has_arm_fhm() ? "yes" : "no");
```
Comment (Maratyszcza) [inline tools/isa-info.c:163]: Move one line up to group with "FP16 arithmetics" which is closely related

**Review by Maratyszcza — CHANGES_REQUESTED** [review]

**Comment by kulinseth** [comment]
@Maratyszcza addressed your comments. Thanks.

**Comment by kulinseth** [comment]
@malfet , 
```
/home/runner/work/cpuinfo/cpuinfo/src/arm/linux/cpuinfo.c:291:29: error: use of undeclared identifier 'CPUINFO_ARM_LINUX_FEATURE_ASIMDFHM'
[86](https://github.com/pytorch/cpuinfo/runs/7395025124?check_suite_focus=true#step:6:87)
                                        processor->features |= CPUINFO_ARM_LINUX_FEATURE_ASIMDFHM;
[87](https://github.com/pytorch/cpuinfo/runs/7395025124?check_suite_focus=true#step:6:88)
                                                               ^
[88](https://github.com/pytorch/cpuinfo/runs/7395025124?check_suite_focus=true#step:6:89)
1 error generated.
[89](https://github.com/pytorch/cpuinfo/runs/7395025124?check_suite_focus=true#step:6:90)
[7/239] Building C object CMakeFiles/cpuinfo.dir/src/arm/linux/init.c.o
[90](https://github.com/pytorch/cpuinfo/runs/7395025124?check_suite_focus=true#step:6:91)
ninja: build stopped: subcommand failed.
[91](https://github.com/pytorch/cpuinfo/runs/7395025124?check_suite_focus=true#step:6:92)
Error: Process completed with exit code 1.
```
This looks like genuine failure. Do we need to make this Mac arm specific ?

File: src/arm/linux/cpuinfo.c [hunk 6]
```diff
@@ -287,6 +287,8 @@ static void parse_features(
 					#if CPUINFO_ARCH_ARM64
 						processor->features |= CPUINFO_ARM_LINUX_FEATURE_ASIMDRDM;
 					#endif
+				} else if (memcmp(feature_start, "asimdfhm", feature_length) == 0) {
+					processor->features |= CPUINFO_ARM_LINUX_FEATURE_ASIMDFHM;
```
Comment (Maratyszcza) [inline src/arm/linux/cpuinfo.c:291]: Wrap in `#if CPUINFO_ARCH_ARM64`

**Review by Maratyszcza — CHANGES_REQUESTED** [review]

File: src/arm/linux/aarch64-isa.c [hunk 7]
```diff
@@ -138,5 +138,12 @@ void cpuinfo_arm64_linux_decode_isa_from_proc_cpuinfo(
 	if (features2 & (CPUINFO_ARM_LINUX_FEATURE2_BF16 | CPUINFO_ARM_LINUX_FEATURE2_SVEBF16)) {
 		isa->bf16 = true;
 	}
+	if (features2 & CPUINFO_ARM_LINUX_FEATURE2_SVEBF16) {
```
[the remark this answers is not in the document]
Comment (kulinseth) [reply src/arm/linux/aarch64-isa.c:141]: I see, I locally build and I am not seeing build failure when I run the `sh scripts/local-build.sh` and then I test with `build/bin/init-test` binary.

[the remark this answers is not in the document]
Comment (kulinseth) [reply src/arm/linux/aarch64-isa.c:141]: Regardless I removed the line.

**Comment by kulinseth** [comment]
> Fails on AArch32 because identifier is only defined on AArch64

Thanks, removed that line.

File: src/arm/linux/cpuinfo.c [hunk 8]
```diff
@@ -287,6 +287,10 @@ static void parse_features(
 					#if CPUINFO_ARCH_ARM64
 						processor->features |= CPUINFO_ARM_LINUX_FEATURE_ASIMDRDM;
 					#endif
+				} else if (memcmp(feature_start, "asimdfhm", feature_length) == 0) {
+					#if CPUINFO_ARCH_ARM64
+					processor->features |= CPUINFO_ARM_LINUX_FEATURE_ASIMDFHM;
```
Comment (Maratyszcza) [inline src/arm/linux/cpuinfo.c:292]: Indent the line as if it was within a normal `if` block

File: src/arm/linux/aarch64-isa.c [hunk 9]
```diff
@@ -138,5 +138,9 @@ void cpuinfo_arm64_linux_decode_isa_from_proc_cpuinfo(
 	if (features2 & (CPUINFO_ARM_LINUX_FEATURE2_BF16 | CPUINFO_ARM_LINUX_FEATURE2_SVEBF16)) {
 		isa->bf16 = true;
 	}
+	if (features & CPUINFO_ARM_LINUX_FEATURE_ASIMDFHM) {
+		isa->fhm = true;
+	}
+
```
Comment (Maratyszcza) [inline src/arm/linux/aarch64-isa.c:144]: Remove the extra newline

**Review by Maratyszcza — CHANGES_REQUESTED** [review]
Thanks! Last formatting changes, and it is good to go

**Review by Maratyszcza — APPROVED** [review]

**Pull request merged**